### ISSUE TITLE
Profiles in Jakarta EE

### ISSUE THREAD

**Issue by ivargrimstad** [eclipse-ee4j/jakartaee-platform#1]
Define if and how the concept of profiles should be used in regard to Jakarta EE.
Continue with the Full and Web profiles? Add more profiles? Abandon the profiles concept?

The roadmap should contain some direction in this regard.

**Comment by keilw** [comment]
That's an interesting question. Some of us with a long history in the Java EE EG demanded more than just the Full and Web profile many years ago. The whole idea and name "MicroProfile" was pretty much born out of that ;-) If Jakarta EE shall be a good platform for Microservices or similar use cases, then adding more and smaller/versatile profiles is inevitable. Or one could drop the idea of profiles altogether, as long as compatibility claims can still be made in a reliable way. Defining just the one or two TCK tests for a particular provider so that one states "We are Jakarta EE compatible" will be a challenge either way.

**Comment by mkarg** [comment]
+1 for finer grained feature separation and more profiles collecting them for typical usage scenarios

**Comment by bshannon** [comment]
I've talked about this before but let me add my perspective on profiles...

Profiles are solving two problems:

- The platform is too big!  I don't use all that stuff!
- The platform is too big!  I don't want to implement all that stuff!

The first is a developer issue, the second is a vendor issue.

Would adding more profiles would encourage additional vendors to produce
Jakarta EE branded products?  Not just more and different repackagings of
the same open source components, but more completely new implementations?
There were *many* completely independent implementations of J2EE in the
early days, but consolidation and competition has reduced the number
significantly.  Would we get a lot more Jakarta EE branded products if
we defined more profiles?  I'm skeptical, but it's worth considering.

I think the developer issue is best solved by modularity, and that was
our plan for Java EE 9, building on the module support introduced in
Java SE 9.  Developers want to have the choice of all these capabilities
that they can use when needed, they just don't want to be forced to
"pay" for them, in download size, memory size, runtime overhead, etc.
So, just like Java SE has moved away from profiles and has embraced
modules, I think Jakarta EE should do the same.

Profiles solve the vendor problem, modules solve the developer problem.

**Comment by mkarg** [comment]
I thnk profile solve one more developer problem: It is rather comfortable and safe. If you are a Jakarta EE charter, there are lots of APIs to understand, with possibly lots more of modules you need to understand what they do for you. But it is very easy to understand that you just need one single dependency (on the profile) to always have what you need e. g. for a web application or for a cloud application. You can rely on this always working, without fully understanding all the modules implied. So without profiles learning Jakarta EE will become harder.

**Comment by kazumura** [comment]
Adding too many profiles makes the market confuse and reduces the vendor compatibilities.
Assume that a vendor provides X-Profile product and another vendor provides Y-Profile product,
both products are Jakarta EE compatible,
but X-Profile applications and Y-Profile applications are not compatible.
More profiles will be less compatible, that is, Jakarta EE brand value.
I think the number of profiles should be at most 3 or 4.

**Comment by keilw** [comment]
MicroProfile already got both vendors and users into a "Compatibility Hell". Every vendor can pretty much pick what they like and are able to handle. Except some of the basic underlying standards (which are Jakarta EE "Micro Profile" if you like;-) every project can claim compatibility even if they just support one or a few features.

**Comment by bshannon** [comment]
@mkarg "Learnability" is definitely an issue to be considered.  But have you looked at all the APIs in Java SE?  Learnability is an issue there too, and yet they haven't resorted to profiles.  Good developer documentation and tutorials can help a lot.

Profiles also have the reverse problem.  Once you exceed what's available in a profile, how do you know where to turn next?

With modules, you can start with just a few modules from a vendor, then add more modules from the vendor when your needs grow, knowing that all the modules will work together.

@kazumura I completely agree that too many profiles works against compatibility and reduces the value of the Jakarta EE brand.  3 or 4 profiles seems like a reasonable limit.  Given the full platform, the web profile, and a potential micro profile, I think we're pretty close to the limit.

**Comment by mkarg** [comment]
@bshannon Yes, docs and tutorials are a help. But on the other hand, it might be a good idea to have Java SE profiles to clearly separate the topics. When I started with Java SE back in 1997, I was somewhat confuses seeing all the lots of APIs existing. It might have helped if there would have been just a core set, and addition sets like "working with files", "working with streams" and so on. So I see profiles missing on Java SE, too. Too many profiles is not a good thing, but we should not concentrate on the number *wanted* but on the number *needed*.

**Comment by keilw** [comment]
For Java SE getting modules was hard enough. The "base" module is far too big and contains among other things at least 3 often competing Date/Time libraries (one in Concurrency, the initial java.util.Date and java.time) none of which can be split out well and the one that's based on a JSR has the RI of that JSR baked into the API, so it prevents any independent implementation, too. 
For Jakarta EE there are some that depend on others while some, e.g. JSON Processing are so independent, they even work in a standalone Java SE environment.

**Comment by mkarg** [comment]
This sounds like if Java SE would be monolithic, but in fact it already _has_ profiles: https://www.oracle.com/technetwork/java/embedded/resources/tech/compact-profiles-overview-2157132.html.

**Comment by bshannon** [comment]
@mkarg And just like Java EE profiles, that was viewed as a stopgap until modules were introduced.  Modules and jlink do everything that Java SE profiles did, and more.

**Comment by kazumura** [comment]
The difference between Java SE and Java EE is the target for brandings.
Java SE has only single brand 'Java SE' even if it has a lot of profiles.
On the other hand, the number of Java EE (and Jakarta EE) brandings is eauls to the number of profiles.
We cannnot discuss Java SE and Java EE (Jakarta EE)  same way.

**Comment by mkarg** [comment]
@kazumura I completely do not understand this argument. Can you please explain it a bit more elaborate? In particular the stuff about branding and why we cannot compare SE and EE.

**Comment by mkarg** [comment]
@kazumura I still do not understand what your point is. If we decide for more profiles, we simply have more brands. So what?

**Comment by kazumura** [comment]
@mkarg This is exactly the problem what I pointed out.
<"https://github.com/eclipse-ee4j/jakartaee-platform/issues/1#issuecomment-453462491">
It will make the market confuse and reduce the value of the brand.
Many vendors can claim "Our product is Jakarta EE compatible!",
but actually there may be no compatibility among vendors' products.

**Comment by mkarg** [comment]
This is nonsense. If we decide to set up a certification program for a brand called "Jakarta EE 9 Edge Computing Profile" this will not confuse anybody as long as we clearly define the content of such a program.

**Comment by keilw** [comment]
Any certification is only as good as transparent TCK tests to proof it. Anybody could pay money for some kind of certificate, but they should also be reliable, sustainable and understandable to everyone.

**Comment by mkarg** [comment]
If a "Jakarta EE 9 Edge Computing Profile" is clearly described in terms of tests to be passed I cannot see a risk.

**Comment by mkarg** [comment]
Did anyone actually propose _100_ profiles?

**Comment by kazumura** [comment]
If you agree 100 profiles are problems,
then, we should discuss about the proper number of profiles
in order to maintain the compatibility brand value.

https://github.com/eclipse-ee4j/jakartaee-platform/issues/1#issuecomment-453462491
I said at most 3 or 4.

**Comment by keilw** [comment]
Those working on and using MicroProfile might be close to 100 profiles ;-D
EAP 7.2 supports only 3 or 4 of more than a dozen MP features, thus shows how many of them are Enterprise ready at this point. Many others pick between 1 and several. 

If a "free style" mode to add other features (e.g. MP) at your own risk was on top, then I guess 3, 4 or up to 6 profiles could be sufficient.

**Comment by mkarg** [comment]
Certainly I agree that 100 profiles is a problem. In fact, I doubt that even the MP people want to have 100 profiles. For me it more looks like the MP people produce a bunch of single features that we could choose from to form 3-5 profiles.

**Comment by keilw** [comment]
MP is irrelevant for Jakarta EE because it does not standardize anything. It may use profiles in Jakarta EE and then formulate some kind of "profiles" (or another word, given the whole project is named after the 3rd Profile we asked for in Java EE 6/7) in addition to it.

**Comment by mkarg** [comment]
MP is no "irrelevant". The vendors gathered there are the same gathered in Jakarta EE. It is clear that APIs developed by MP and supported by the Jakarta EE implementations are candidates for inclusion to Jakarta EE. At JAX-RS for example we discuss adoption of MP Config API. We see the activities at MP as an incubator for inclusion into standards like Jakarta EE.

**Comment by keilw** [comment]
As a dependency it is irrelevant for Jakarta EE. MP in a future version will use Jakarta EE not the other way round. If some specs evolve from current MP features (config sounds like an interesting candidate) then MP and its users will gradually have to move from the incubator/infant API to a more mature specification in Jakarta EE. At the moment there is a lot of duplication of efforts and fragmented codebases which make it harder for potential users to adopt them.

**Comment by mkarg** [comment]
@keilw That is true, but what actually do you want to reach by telling this all in *this* thread? What is the point you want to make with it?

**Comment by keilw** [comment]
It was pointing out that outside a real standard you quickly end up with too many "profiles". Not sure, where the whole "100 profiles" discussion tried to go in the first place? ;-)

**Comment by keilw** [comment]
The profiles document in this repository still talks about "Java EE Profiles" everywhere and only mentions building blocks as "Jakarta" specs. Is that intentional or will that content not even be part of Jakarta EE 8 yet?
